Replace vim.fn/vim.env in vim.fs

mike325 opened the pull request:
`vim.fs` functions like, `vim.fs.find`, could be useful in fast event contexts and inside uv threads,

Ex. Looking for files/dirs in slow FS or traversing deep repositories can hang the ui and can be delegated to background threads 

```lua
local function find_cpp_src(path)
    local fs = require 'vim.fs'

    local function filter(filename)
        return filename:match '.+%.cpp$' ~= nil
    end

    return vim.json.encode(fs.find(filter, { limit = math.huge, path = path }))
end

local function cb(results)
    if type(results) == type '' then
        vim.g.cpp_files = vim.json.decode(results)
    end
end

local work = vim.loop.new_work(find_cpp_src, cb)
work:queue '~/foo/'
```

Right now this is partially possible initializing `vim.env` and not using `basename/dirname` functions which include `vim.fs.find` going in upward direction

lewis6991 commented:
This will regress #17501

gpanders commented:
Eliminating `vim.fn` would be good I think, for the reasons you mentioned. The concern for me is that if we roll our own implementation then we have to be very careful about catching all of the corner cases that `fnamemodify` has already learned to handle over the years. We should be able to address this with a thorough suite of unit tests, however.

zeertzjq commented:
> This will regress #17501

The change is only is `fs.lua`.

zeertzjq commented:
BTW existing tests haven't passed yet.

Mike325 commented:
> Just to note, we can now mark functions in `vim.fn` to be allowed to run in fast as long as the said functions don't alter buffer state.

Are `vim.fn`'s that don't alter the buffer's state sharable between threads too ?

lewis6991 commented:
Not currently.

Mike325 commented:
I could start writing some test to compare the outputs of `fnamemodify` basename/dirname with the new lua implementations , I don't think getting rid of `vim.env` could cause any side effect since replacing it with luv implementation should behave just as the internal `setenv()` vim function.

mike325 committed:
fix: avoid vim.fn and vim.env in vim.fs module

mike325 committed:
test: add vim.fs.basename/dirname tests

check equivalence of vim.fnamemodify with vim.fs.basename/dirname

mike325 committed:
fix: add more fs tests and fix vim.fs.basename empty string handling

gpanders commented:
Windows test failure is relevant.

mike325 committed:
test: add more windows test cases

Mike325 commented:
Yeah, I see, unfortunately I don't have a windows machine to test the changes, I'll try to spin up a VM to test the functions

mike325 committed:
ci: fix missing windows tests cases

mike325 committed:
ci: add local-test windows detection

Mike325 commented:
After latest changes, Windows and MacOS are still failing but I'm not sure the failures are related to my changes, I had no problems testing locally using `TEST_FILTER='vim.fs.*' make functionaltest` in windows and linux (missing macos tho)

theHamsta commented on runtime/lua/vim/fs.lua:
```diff
@@ -1,5 +1,7 @@
 local M = {}
 
+local iswin = package.config:sub(1, 1) == '\\'
```
perhaps `vim.loop.os_uname` would document itself better

mike325 committed:
feat: replace vim.fs.basename/dirname windows detection method

Replace `package.config:sub(1, 1) == '\\'` for `vim.loop.os_uname().sysname == 'Windows_NT'`

gpanders commented:
> After latest changes, Windows and MacOS are still failing but I'm not sure the failures are related to my changes, I had no problems testing locally using `TEST_FILTER='vim.fs.*' make functionaltest` in windows and linux (missing macos tho)

Neither macOS or Windows failures look relevant to this PR.

gpanders commented on test/functional/lua/fs_spec.lua:
```diff
@@ -70,7 +70,7 @@ describe('vim.fs', function()
   describe('dirname()', function()
     it('works', function()
-      local test_windows = package.config:sub(1, 1) == '\\'
+      local test_windows = vim.loop.os_uname().sysname == 'Windows_NT'
```
In the tests we have a helper function `is_os`. Simply add

```lua
local is_os = helpers.is_os
```

to the top of the file (if it's not already there) and use it like this:

```lua
is_os('win')
```

mike325 committed:
test: use is_os helper in vim.fs.basename/dirname tests

gpanders commented on runtime/lua/vim/fs.lua:
```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
   end
-  return vim.fn.fnamemodify(file, ':h')
+  vim.validate({ file = { file, 's' } })
+  if file:match('^%w:[\\/]?$') and iswin then
```
Probably is only a minor consideration, but maybe put `iswin` in the condition first, so that it fails quickly when not on Windows (otherwise, it will run `string.match()` *every* time, even if `iswin` is false).

gpanders commented on runtime/lua/vim/fs.lua:
```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
+    return file:gsub('\\', '/')
+  elseif not file:match('[\\/]') then
+    return '.'
+  elseif file:match('^/[^/]+$') or file == '/' then
```
Similar comment here, it's cheaper/easier to check `file == '/'` so put that in the condition first so that we don't call `string.match()` if we don't need to.

mike325 committed:
fix: invert if conditions and fix double gsub() return value

Check simple condition first
Surround gsub return to prevent from returning the substituted index

dundargoc commented on runtime/lua/vim/fs.lua:
```diff
@@ -1,5 +1,7 @@
 local M = {}
 
+local iswin = vim.loop.os_uname().sysname == 'Windows_NT'
```
Don't create custom os detection please, these things tend to proliferate.

```
local iswin = is_os('win')
```

is fine.

gpanders replied:
Do we have `is_os` outside of the test suite?

dundargoc replied:
I can't read. I missed that this was outside of test. Ignore.

Mike325 commented:
@dundargoc do you mean the conditions to match the cases are difficult to read ?

dundargoc commented:
@Mike325 I needed to give a reason as to why I wanted to dismiss my review, since I realized it was a mistake. I wrote that to myself as a joke. Don't worry about it, the PR is fine.
